R/mwrite.R

```diff
@@ -0,0 +1,341 @@
+#' file. `mread_yaml()` does not write model settings into the file but rather 
+#' update the model object directly with data read back from the `yaml` file.
+#' 
+#' @examples
+#' mod <- house()
+#' 
+#' temp <- tempfile(fileext = ".yaml")
+#' 
+#' mwrite_yaml(mod, file = temp)
+#' 
+#' # Note: this model is not compiled
+#' mod <- mread_yaml(temp, model = "new-house", compile = FALSE)
+#' mod
+#' 
+#' cppfile <- yaml_to_cpp(temp)
+#' 
+#' readLines(cppfile)
+#' 
+#' @return 
+#' A model object. 
+#' 
+#' @seealso [mwrite_yaml()], [yaml_to_cpp()]
+#' 
+#' @md
+#' @export
+mread_yaml <- function(file, model = basename(file), project = tempdir(), ...) {
+  x <- mwrite_parse_yaml(file)
+  parsed_to_model(x, model, project, ...)
+}
+
+mwrite_parse_yaml <- function(file) {
+  require_yaml()
+  text <- readLines(file)
+  l <- yaml::yaml.load(text)
+  if(!identical(l$source, "mrgsolve::mwrite")) {
+    abort("the yaml source file was not written by `mwrite_yaml()`.")  
+  }
+  l
+}
+
+#' @rdname mread_yaml
+#' @export
+yaml_to_cpp <- function(file, model = basename(file), project = getwd(), 
+                        update = TRUE) {
+  x <- mwrite_parse_yaml(file)
+  x <- parsed_to_cppfile(x, model, project, update)
+  invisible(x$cppfile)
+}
+
+# Take in content parsed from yaml file, clean up, write to cpp file
+# @return a cleaned-up version of x with `cppfile` slot added
+parsed_to_cppfile <- function(x, model, project, update = FALSE) {
+  prob <- character(0)
+  if(sum(nchar(x$prob))) {
+    prob <- c("$PROB", x$prob, "")  
+  }
+  
+  param <- character(0)
+  if(length(x$param)) {
+    param <- c("$PARAM", tocode(x$param), "")
+  }
+  init <- character(0)
+  if(length(x$init)) {
+    init <- c("$INIT", tocode(x$init), "")
+  }
+  capture <- character(0)
+  if(length(x$capture)) {
+    capture <- c("$CAPTURE", x$capture, "")
+  }
+  
+  x$update$add <- as.numeric(x$update$add)
+  
+  omega <- character(0)
+  if(length(x$omega$data)) {
+    x$omega$labels <- lapply(x$omega$labels, as.character)
+    x$omega$names <- lapply(x$omega$names, as.character)
+  }
+  for(i in seq_along(x$omega$data)) {
+    header <- "@block"
+    if(any(x$omega$labels[[i]] != "...")) {
+      o_labels <- paste0(x$omega$labels[[i]], collapse = " ")
+      header <- c(header, paste0("@labels ", o_labels))
+    }
+    omega <- c(omega, "$OMEGA", header, x$omega$data[[i]], "")  
+  }
+  
+  sigma <- character(0)
+  if(length(x$sigma$data)) {
+    x$sigma$labels <- lapply(x$sigma$labels, as.character)
+    x$sigma$names <- lapply(x$sigma$names, as.character)
+  }
+  for(i in seq_along(x$sigma$data)) {
+    header <- "@block"
+    if(any(x$sigma$labels[[i]] != "...")) {
+      s_labels <- paste0(x$sigma$labels[[i]], collapse = " ")
+      header <- c(header, paste0("@labels ", s_labels))
+    }
+    sigma <- c(sigma, "$SIGMA", header, x$sigma$data[[i]], "")  
+  }
+  
+  code <- c(prob, param, init, omega, sigma, x$code, capture)
```

Comment:
Capture information pushed to the end of the control stream; handling this explicitly now, like param and init.